### PR TITLE
Feature/convert java tocpp mobility response

### PR DESCRIPTION
**Pull request by Anishdeva**
# PR Details
This branch adds Mobility Response Message to the cpp message node. 
## Description
To support the upgrade to ROS1 Noetic and ROS2 all java code must be removed from the system. This story is part of a set of stories to convert the java message package to C++ and integrate its logic with the newer message_cpp package. The scope of this story is to specifically port the MobilityResponse message functionality from the Java Message node to the message_cpp node, leaving all other messages in the Java node untouched. Follow-on stories will be created to convert each message type still existing in the Java Message node.

Mobility Response subscriber and publisher was added to the message consumer node which eventually supposed to run the encoding and decoding for all message types. A header and cpp file for MobilityResponse was added, which has the specific encoding and decoding functions.
The Mobility Response node call in java was changed to return null.
## Motivation and Context

Required for future upgrade to ROS Noetic.
## How Has This Been Tested?
Unit tests were created for encoding, the binary array generated during the encoding is used for a decoding test.

Unit test values were taken from the java code tests, since those tests did not contain values for the boolean is_accepted, a value of 1 was assumed. The binary array generated after encoding was printed and used as input to the decoding function. The output from decoding was checked against values that were given as input during encoding.

## Types of changes

- [ ] Defect fix (non-breaking change that fixes an issue)
- [ ] New feature (non-breaking change that adds functionality)
- [ ] Breaking change (fix or feature that cause existing functionality to change)

## Checklist:

- [ ] I have added any new packages to the sonar-scanner.properties file
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the **CONTRIBUTING** document.
[CARMA Contributing Guide](https://github.com/usdot-fhwa-stol/carma-platform/blob/develop/Contributing.md) 
- [x] I have added tests to cover my changes.
- [x] All new and existing tests passed.

### CODE REVIEW

File: carma-messenger-core/cpp_message/src/MessageConsumer.cpp [hunk 1]
```diff
@@ -0,0 +1,102 @@
...
+        mobility_response_message_sub_=nh_->subscribe<>("outgoing_plain_mobility_operation_message",5, &MessageConsumer::outbound_mobility_response_message_callback,this);
+    }
+    
+    void MessageConsumer::inbound_binary_callback(const cav_msgs::ByteArrayConstPtr& msg)
+    {   //find message type and then send for decoding
+        if(msg->messageType=="MobilityOperation")   
+        {
+            std::vector<uint8_t> array=msg->content;
+            Mobility_Operation decode;
+            auto output=decode.decode_mobility_operation_message(array);
```
Comment (kjrush) [inline carma-messenger-core/cpp_message/src/MessageConsumer.cpp:46]: Is there a reason to initialize this output variable if it serves no purpose? It appears to be allocated, but never used and then falls out of scope.

Comment (kjrush) [reply carma-messenger-core/cpp_message/src/MessageConsumer.cpp:46]: I'm also not seeing where this data flows to after it is decoded? This `Mobility_Operation decode` object also appears to die on the stack here after it's decoded. I'd think it should be converted into a ROS message `cav_msgs/MobilityOperation.msg` and then published on that `incoming_mobility_messsage_decoded` topic.

File: carma-messenger-core/cpp_message/src/MessageConsumer.cpp [hunk 2]
```diff
@@ -0,0 +1,102 @@
...
+
+    void MessageConsumer::outbound_mobility_operation_message_callback(const cav_msgs::MobilityOperation& msg)
+    {//encode and publish as outbound binary message
+        Mobility_Operation encode;
+        auto res=encode.encode_mobility_operation_message(msg);
+        if(res)
+        {
+            //copy to byte array msg
+            cav_msgs::ByteArray output;
+            output.content=res.get();
```
Comment (kjrush) [inline carma-messenger-core/cpp_message/src/MessageConsumer.cpp:75]: Does this work as expected? I'm always a little bit iffy about moving arrays around using `=`, but if `output.content` is a pointer then this seems okay.

File: carma-messenger-core/cpp_message/src/MessageConsumer.cpp [hunk 3]
```diff
@@ -0,0 +1,102 @@
...
+
+    void MessageConsumer::outbound_mobility_response_message_callback(const cav_msgs::MobilityResponse& msg)
+        {//encode and publish as outbound binary message
+        Mobility_Response encode;
+        auto res=encode.encode_mobility_response_message(msg);
+        if(res)
+        {
+            //copy to byte array msg
+            cav_msgs::ByteArray output;
+            output.content=res.get();
```
Comment (kjrush) [inline carma-messenger-core/cpp_message/src/MessageConsumer.cpp:93]: Same thoughts as above.

File: carma-messenger-core/cpp_message/src/MobilityOperation_Message.cpp [hunk 4]
```diff
@@ -0,0 +1,257 @@
...
+
+namespace Message_cpp
+{
+    boost::optional<cav_msgs::MobilityOperation> Mobility_Operation::decode_mobility_operation_message(std::vector<uint8_t>& binary_array){
+        
+        cav_msgs::MobilityHeader header;
+        cav_msgs::MobilityOperation output;
+        //decode results - stored in binary_array
+        asn_dec_rval_t rval;
+        MessageFrame_t* message=0;
```
Comment (kjrush) [inline carma-messenger-core/cpp_message/src/MobilityOperation_Message.cpp:31]: For clarity's sake, I'd recommend using `nullptr` here instead of the numerical `0`

File: carma-messenger-core/cpp_message/src/MobilityOperation_Message.cpp [hunk 5]
```diff
@@ -0,0 +1,257 @@
...
+           
+            //convert strategy from char array to string (TestMessage03 for MobilityOperation)
+            std::string sender_id, recipient_id, sender_bsm_id, plan_id,timestamp_string, strategy,strategy_params;
+            uint64_t timestamp;
+            //get sender id
+            auto str_len=message->value.choice.TestMessage03.header.hostStaticId.size;
+            if(str_len<=STATIC_ID_MAX_LENGTH && str_len!=0)
+            {
+                for(auto i=0;i<str_len;i++){
+                    sender_id +=message->value.choice.TestMessage03.header.hostStaticId.buf[i];
```
Comment (kjrush) [inline carma-messenger-core/cpp_message/src/MobilityOperation_Message.cpp:55]: Is this correct to how the original implementation was? I'm trying to wrap my head around the implementation here. A static ID like this is a 4-byte value (if I recall correctly), adding the 4 bytes together individually != the original 4 byte string: (e.g. 11001010<sub>2</sub> = 202<sub>10</sub>, 1100<sub>2</sub> + 1010<sub>2</sub> = 10110<sub>2</sub> or 22<sub>10</sub>).

To correctly do this we'd need to bitshift and add: `1100 << 8 + 1010 = 11001010`.

If this is how the Java code implemented it, we can think on that this may be a load-bearing defect.

File: carma-messenger-core/cpp_message/src/MobilityOperation_Message.cpp [hunk 6]
```diff
@@ -0,0 +1,257 @@
...
+                sender_id=STRING_DEFAULT;
+            }
+            header.sender_id=sender_id;
+
+            //get recepient id
+            str_len=message->value.choice.TestMessage03.header.targetStaticId.size;
+            if(str_len<=STATIC_ID_MAX_LENGTH && str_len!=0)
+            {
+                for(auto i=0;i<str_len;i++){
+                    recipient_id +=message->value.choice.TestMessage03.header.targetStaticId.buf[i];
```
Comment (kjrush) [inline carma-messenger-core/cpp_message/src/MobilityOperation_Message.cpp:69]: Similar concern as above, worth double checking.

File: carma-messenger-core/cpp_message/src/MobilityOperation_Message.cpp [hunk 7]
```diff
@@ -0,0 +1,257 @@
...
+            }
+
+            header.recipient_id=recipient_id;
+            
+            //get bsm id
+            str_len=message->value.choice.TestMessage03.header.hostBSMId.size;
+            if(str_len==BSM_ID_LENGTH)
+            {
+                for(auto i=0;i<str_len;i++){
+                    sender_bsm_id +=message->value.choice.TestMessage03.header.hostBSMId.buf[i];
```
Comment (kjrush) [inline carma-messenger-core/cpp_message/src/MobilityOperation_Message.cpp:84]: Same as above, pretty much anywhere we have a copy of multiple bytes.

File: carma-messenger-core/cpp_message/src/MobilityOperation_Message.cpp [hunk 8]
```diff
@@ -0,0 +1,257 @@
...
+        {
+            string_content_params[i]=plainMessage.strategy_params[i];
+        }
+        message->value.choice.TestMessage03.body.operationParams.buf=string_content_params;
+        message->value.choice.TestMessage03.body.operationParams.size=string_size;
+        
+        //encode message
+        ec=uper_encode_to_buffer(&asn_DEF_MessageFrame, 0 , message , buffer , buffer_size);
+         
+        //log a warning if that fails
```
Comment (kjrush) [inline carma-messenger-core/cpp_message/src/MobilityOperation_Message.cpp:244]: Mention of logging a warning, but no actual log statement.

File: carma-messenger-core/cpp_message/src/MobilityResponse_Message.cpp [hunk 9]
```diff
@@ -0,0 +1,231 @@
+/*
```
Comment (kjrush) [inline carma-messenger-core/cpp_message/src/MobilityResponse_Message.cpp:1]: I thought the response stuff would be in a separate PR, is this intended to be here?

File: carma-messenger-core/cpp_message/src/main.cpp [hunk 10]
```diff
@@ -23,5 +24,7 @@ int main(int argc, char**argv)
 {
     ros::init(argc,argv,"cpp_message_node");
     cpp_message::Message node;
+    Message_cpp::MessageConsumer node_2;
```
Comment (kjrush) [inline carma-messenger-core/cpp_message/src/main.cpp:27]: I think this will prevent `node.run()` from happening, as `node_2.run()` blocks while the node is running. These MobilityMessage callbacks/etc should probably just be integrated into the existing node to that it all runs as a single unit. We can discuss if you want.

File: carma-messenger-core/message/.gradle/2.14.1/taskArtifacts/cache.properties [hunk 11]
```diff
@@ -0,0 +1 @@
+#Fri Jul 24 06:57:58 PDT 2020
```
Comment (kjrush) [inline carma-messenger-core/message/.gradle/2.14.1/taskArtifacts/cache.properties:1]: Let's also make sure all these build artifacts are cleaned up as well, only source code should be committed.

File: carma-messenger-core/message/message/build/install/message/bin/message [hunk 12]
```diff
@@ -0,0 +1,164 @@
+#!/usr/bin/env bash
```
Comment (kjrush) [inline carma-messenger-core/message/message/build/install/message/bin/message:1]: Same as above, only the `src` folder of the workspace should be committed. Though, seeing as this one is under carma-messenger-core, maybe this is an issue with the way the repo is setup or your local workspace is setup. Not sure how this is here.

File: carma-messenger-core/message/message/src/main/java/gov/dot/fhwa/saxton/carma/message/factory/DSRCMessageFactory.java [hunk 13]
```diff
@@ -31,9 +31,11 @@
         case "MobilityPath":
             return new MobilityPathMessage(log, factory);
         case "MobilityResponse":
-            return new MobilityResponseMessage(log, factory);
+            return null;
```
Comment (kjrush) [inline carma-messenger-core/message/message/src/main/java/gov/dot/fhwa/saxton/carma/message/factory/DSRCMessageFactory.java:34]: May be worth also making sure this node doesn't actually receive these message (by removing or changing the topic mappings that it subscribes to, as I could see this causing a NullPointerException at runtime.

File: carma-messenger-core/cpp_message/include/MobilityOperation_Message.h [hunk 14]
```diff
@@ -0,0 +1,39 @@
+/*
```
[the remark this answers is not in the document]
Comment (Anishdeva) [reply carma-messenger-core/cpp_message/include/MobilityOperation_Message.h:1]: Line 1 here is #pragma once

File: carma-messenger-core/cpp_message/launch/MobilityOperation_message.launch [hunk 15]
```diff
@@ -0,0 +1,18 @@
+<?xml version="1.0"?>
```
[the remark this answers is not in the document]
Comment (Anishdeva) [reply carma-messenger-core/cpp_message/launch/MobilityOperation_message.launch:1]: Thanks for pointing this out, I moved all messages under cpp node in the CMakeLists, so don't need a separate launch file.

File: carma-messenger-core/cpp_message/src/MessageConsumer.cpp (hunk 1 above)
Comment (Anishdeva) [reply carma-messenger-core/cpp_message/src/MessageConsumer.cpp:46]: Yes, you're right, I'm publishing it now. Output is a mobilityOperations decoded ros message, it needs to published.

File: carma-messenger-core/cpp_message/src/MobilityResponse_Message.cpp (hunk 9 above)
Comment (Anishdeva) [reply carma-messenger-core/cpp_message/src/MobilityResponse_Message.cpp:1]: This one is the MobilityResponse PR. It builds on top of the MobilityOperation branch which is why it includes both.

File: carma-messenger-core/cpp_message/src/main.cpp (hunk 10 above)
Comment (Anishdeva) [reply carma-messenger-core/cpp_message/src/main.cpp:27]: Changed this, geofence node now launches from inside the MessageConsumer run function.

File: carma-messenger-core/cpp_message/include/MobilityOperation_Message.h (hunk 14 above)
[the remark this answers is not in the document]
Comment (Anishdeva) [reply carma-messenger-core/cpp_message/include/MobilityOperation_Message.h:1]: Added pragma once to all header files.

File: carma-messenger-core/cpp_message/include/MobilityOperation_Message.h [hunk 16]
```diff
@@ -0,0 +1,39 @@
...
+ * the License.
+ */
+#include "Messages.h"
+
+namespace Message_cpp
+{
+    class Mobility_Operation
+    {
+        private:
+        //constants
```
[the remark this answers is not in the document]
Comment (Anishdeva) [reply carma-messenger-core/cpp_message/include/MobilityOperation_Message.h:23]: Made suggested changes.

File: carma-messenger-core/cpp_message/include/MobilityOperation_Message.h [hunk 17]
```diff
@@ -0,0 +1,39 @@
...
+            std::string STRING_DEFAULT="[]";
+            int TIMESTAMP_LENGTH=std::to_string(INT64_MAX).length();
+            std::string GUID_DEFAULT= "00000000-0000-0000-0000-000000000000";
+            int GUID_LENGTH=GUID_DEFAULT.length();
+            int STRATEGY_MAX_LENGTH=50;
+            int STRATEGY_PARAMS_MAX_LENGTH=1000;
+        
+        public:
+        //helper functions for message decode/encode
+        boost::optional<cav_msgs::MobilityOperation>decode_mobility_operation_message(std::vector<uint8_t>& binary_array);
```
[the remark this answers is not in the document]
Comment (Anishdeva) [reply carma-messenger-core/cpp_message/include/MobilityOperation_Message.h:36]: Added Doxygen format comments to all public functions.

File: carma-messenger-core/cpp_message/src/MobilityOperation_Message.cpp (hunk 8 above)
Comment (Anishdeva) [reply carma-messenger-core/cpp_message/src/MobilityOperation_Message.cpp:244]: Added warning

File: carma-messenger-core/cpp_message/src/MessageConsumer.cpp [hunk 18]
```diff
@@ -60,6 +63,8 @@ namespace Message_cpp
     
     int MessageConsumer::run(){
         initialize();
+        cpp_message::Message geofence_node;
+        geofence_node.run();
```
Comment (kjrush) [inline carma-messenger-core/cpp_message/src/MessageConsumer.cpp:67]: I'm not 100% sure this works as expected, I'm concerned the current way still only gets one (geofence vs other messages) portion of this code running at a time. Putting `initialize()` before calling geofence node may help, but it still seems kind of wonky. Let's discuss a bit more on this if you have time.
